django/core/management/commands/makemigrations.py

```diff
@@ -236,11 +249,98 @@ def handle(self, *app_labels, **options):
+            leaf_migration = loader.graph.nodes[leaf_migration_nodes[0]]
+
+            if leaf_migration.replaces:
+                raise CommandError(
+                    f"Cannot update squash migration {leaf_migration_nodes[0]}."
+                )
+
+            # Stop here if the migration is already applied.
+            if leaf_migration_nodes[0] in loader.applied_migrations:
+                deps_from_same_app = [
+                    dep for dep in leaf_migration.dependencies if dep[0] == app_label
+                ]
+                if deps_from_same_app:
+                    prev_migration_name = loader.graph.nodes[deps_from_same_app[0]].name
+                else:
+                    prev_migration_name = "zero"
+                raise CommandError(
+                    f"Last migration for '{app_label}' is already applied. Either:\n"
+                    "  - Revert the migration by running "
+                    f"'python manage.py migrate {app_label} {prev_migration_name}'\n"
+                    "  - Create a new migration by running "
+                    f"'python manage.py makemigrations {app_label}'"
+                )
+
+            # Build new migration.
+            for migration in app_migrations:
+                leaf_migration.operations.extend(migration.operations)
+
+                for dependency in migration.dependencies:
+                    if isinstance(dependency, SwappableTuple):
+                        if settings.AUTH_USER_MODEL == dependency.setting:
+                            leaf_migration.dependencies.add(
+                                ("__setting__", "AUTH_USER_MODEL")
+                            )
+                        else:
+                            leaf_migration.dependencies.add(dependency)
+                    elif dependency[0] != migration.app_label:
+                        leaf_migration.dependencies.add(dependency)
+
+            # Optimize migration.
+            optimizer = MigrationOptimizer()
+            leaf_migration.operations = optimizer.optimize(
+                leaf_migration.operations, app_label
+            )
+
+            # Update name.
+            previous_migration_path = MigrationWriter(leaf_migration).path
+            suggested_name = (
+                leaf_migration.name[:4] + "_" + leaf_migration.suggest_name()
+            )
+            if leaf_migration.name == suggested_name:
+                new_name = leaf_migration.name + "_updated"
+            else:
+                new_name = suggested_name
+            leaf_migration.name = new_name
+
+            # Check for circular dependencies.
+            try:
+                loader.graph.ensure_not_cyclic()
+            except CircularDependencyError as ex:
+                raise CommandError(
+                    "Cannot update last migration because it would generate a circular "
+                    f"dependency with {ex}."
+                )
```

Comment:
This is untested.

Reply:
With the current methods (`MigrationLoader.build_graph`), I'm not sure we can easily check if we introduced circular dependencies without having actually written the migrations to disk.

I removed this condition for now, since I'm not sure anymore the new `--update` option is the right place for it 🤔 
What do you think @felixxm?

Reply:
IMO, it's unnecessary.